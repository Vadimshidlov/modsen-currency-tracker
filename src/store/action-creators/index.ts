export { getMapData } from "@/store/action-creators/getMapData";
export { getCurrency } from "@/store/action-creators/getCurrency";
export {
    selectSecondCurrency,
    closeModalWindow,
    openModalWindow,
} from "@/store/action-creators/openModalWindow";
export { switchLightTheme, switchDarkTheme } from "@/store/action-creators/toggleTheme";
export {
    errorAuthUserLocation,
    successAuthUserLocation,
} from "@/store/action-creators/authUserLocation";
