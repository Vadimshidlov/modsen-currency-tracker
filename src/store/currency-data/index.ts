export { mockMapData } from "@/store/currency-data/mockMapData";
export { mockBanksApiData } from "@/store/currency-data/mockBanksApiData";
export { mockStockData } from "@/store/currency-data/mockStockData";
