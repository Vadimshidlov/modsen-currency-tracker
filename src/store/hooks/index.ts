export { useActions } from "@/store/hooks/useActions";
export { useTypedSelectorHook } from "@/store/hooks/useTypedSelector";
