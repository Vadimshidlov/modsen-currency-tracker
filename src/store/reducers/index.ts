import { combineReducers } from "redux";
import themeReducer from "@/store/reducers/themeReducer";
import latestCurrencyReducer from "@/store/reducers/latestCurrencyReducer";
import modalWindowReducer from "@/store/reducers/modalWindowReducer";
import mapDataReducer from "@/store/reducers/mapDataReducer";
import userLocationReducer from "@/store/reducers/locationAuthReducer";

export const rootReducer = combineReducers({
    theme: themeReducer,
    currency: latestCurrencyReducer,
    modalWindow: modalWindowReducer,
    mapData: mapDataReducer,
    userLocation: userLocationReducer,
});

export type RootStateType = ReturnType<typeof rootReducer>;
