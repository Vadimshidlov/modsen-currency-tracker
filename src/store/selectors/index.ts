export { selectCurrency, selectModal, selectTheme } from "@/store/selectors/selectors";
