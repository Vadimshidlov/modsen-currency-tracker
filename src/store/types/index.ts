import { BanksDataType } from "@/types";

export type ModalWindowStateType = {
    isOpen: boolean;
    currentCurrencyTitle: string;
    currentCurrencyCode: string;
    currentCurrencyValue: number;
    secondCurrencyTitle: string;
    secondCurrencyCode: string;
    secondCurrencyValue: number;
};

export enum ThemeAction {
    SWITCH_DARK = "SWITCH_DARK",
    SWITCH_LIGHT = "SWITCH_LIGHT",
}

export type OpenWindowPayloadType = {
    currencyData: CurrencyType;
    currentCurrencyTitle: string;
    currencyCode: string;
    currentCurrencyValue: number;
};

export type SecondCurrencyPayloadType = {
    secondCurrencyTitle: string;
    secondCurrencyCode: string;
    secondCurrencyValue: number;
};

export type OpenWindowActionType = {
    type: "OPEN_MODAL_WINDOW";
    payload?: OpenWindowPayloadType;
};

export type SelectSecondCurrencyActionType = {
    type: "SELECT_SECOND_CURRENCY";
    payload?: SecondCurrencyPayloadType;
};

export type CloseWindowActionType = {
    type: "CLOSE_MODAL_WINDOW";
};

export type ModalWindowActionType =
    | OpenWindowActionType
    | CloseWindowActionType
    | SelectSecondCurrencyActionType;

export type ThemeStateType = {
    theme: string | "Dark" | "Light";
};

export type PreviousThemeType = null | string;

export type DarkThemeActionType = {
    type: "SWITCH_DARK";
};

export type LightThemeActionType = {
    type: "SWITCH_LIGHT";
};

export type ThemeActionType = LightThemeActionType | DarkThemeActionType;

export type MapDataStateType = {
    mapData: Partial<BanksDataType>[];
};

export type GetMapDataType = "GET_MAP_DATA";

export type GetMapDataActionType = {
    type: GetMapDataType;
};

export type MapDataActionType = GetMapDataActionType;

export type SuccessLocationAuthType = {
    type: "SUCCES_LOCATION_AUTH";
    payload: {
        userLtt: number;
        userLgt: number;
    };
};

export type UnSuccessLocationAuthType = {
    type: "ERROR_LOCATION_AUTH";
};

export type LocationStateType = {
    isLocationAuth: boolean;
    userLtt: number;
    userLgt: number;
};

export type LocationAuthActionType = SuccessLocationAuthType | UnSuccessLocationAuthType;

export type CurrencyType = {
    meta: MetaType;
    data: CurrencyDataType;
};

export type CurrencyDataType = {
    [currencyCode: string]: {
        code: string;
        value: number;
    };
};

export type CurrentCurrencyStateType = {
    currency: CurrencyType;
    isLoading: boolean;
    error: null | string;
};

export enum CurrentCurrencyAction {
    GET_CURR_CURRENCY = "GET_CURR_CURRENCY",
    GET_CURR_CURRENCY_SUCCES = "GET_CURR_CURRENCY_SUCCES",
    GET_CURR_CURRENCY_ERROR = "GET_CURR_CURRENCY_ERROR",
}

export type GetCurrCurrencyType = {
    type: CurrentCurrencyAction.GET_CURR_CURRENCY;
};

export type GetCurrCurrencySuccesType = {
    type: CurrentCurrencyAction.GET_CURR_CURRENCY_SUCCES;
    payload?: CurrencyType;
};

export type GetCurrCurrencyErrorType = {
    type: CurrentCurrencyAction.GET_CURR_CURRENCY_ERROR;
    payload?: string;
};

export type MetaType = {
    last_updated_at: string;
};
