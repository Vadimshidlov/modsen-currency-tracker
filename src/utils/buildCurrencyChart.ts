import Chart from "chart.js/auto";
import React from "react";
import { getRandomOhlcv } from "@/utils/getRandomOhlcv";

let chart: Chart;

export const buildCurrencyChart = (
    ref: React.RefObject<HTMLCanvasElement>,
    currencyValue: number,
    chartDate: string,
) => {
    // console.log(ref, "ref");
    console.log(currencyValue, chartDate, "buildCurrencyChart");

    return new Promise<void>((resolve, reject) => {
        try {
            const data = {
                datasets: [
                    {
                        // data: [...getRandomOhlcv(1000).datasets[0].data],
                        data: [
                            ...getRandomOhlcv(currencyValue, new Date(chartDate)).datasets[0].data,
                        ],
                        barPercentage: 0.5,
                        barThickness: 6,
                        maxBarThickness: 8,
                        minBarLength: 2,
                    },
                ],
            };

            const ctx = ref.current?.getContext("2d");
            // const ctx = ref.current?.getContext("2d");
            // ctx.clearRect(0, 0, ref.current?.width, ref.current?.height);

            if (ctx) {
                setTimeout(() => {
                    /* if (ref.current.v) {
                        ref.current.chart.destroy();
                    } */

                    // const chart = new Chart(ctx, {
                    if (chart) {
                        // chart.destroy();
                        chart.update();
                    }

                    chart = new Chart(ctx, {
                        type: "candlestick",
                        data,
                        options: {
                            animations: {
                                tension: {
                                    duration: 1000,
                                    easing: "linear",
                                    from: 1,
                                    to: 0,
                                    loop: false,
                                },
                            },
                            responsive: true,
                            tooltips: {
                                mode: "index",
                            },
                            interaction: {
                                intersect: false,
                            },
                            plugins: {
                                title: {
                                    display: true,
                                    text: "Modsen Currency Tracker",
                                },
                                legend: {
                                    display: false,
                                },
                                tooltip: {
                                    intersect: false,
                                    position: "nearest",
                                },
                            },
                            scales: {
                                y: {
                                    display: true,
                                    title: {
                                        display: true,
                                        text: "Value",
                                        position: "bottom",
                                    },
                                    tooltip: {},
                                    position: "right",
                                },
                                xAxis: {
                                    type: "time",
                                    time: {
                                        unit: "day",
                                        tooltipFormat: "YYYY-MM-DD",
                                        displayFormats: {
                                            day: "DD",
                                        },
                                    },
                                    title: {
                                        display: true,
                                        text: "Day",
                                        position: "bottom",
                                    },
                                    barPercentage: 0.5,
                                    barThickness: 1,
                                    grid: {
                                        display: true,
                                        drawOnChartArea: true,
                                        drawTicks: true,
                                    },
                                },
                            },
                        },
                    });
                    // chart.render();
                    // resolve(chart.update());
                    // resolve(chart.render());

                    // chart.render();
                    // chart.render();
                    console.log(chart);

                    // chart.destroy();
                    chart.render();
                    // successChartBuildMessage();
                    resolve();
                }, 1000);
            } else {
                reject(new Error("Could not get canvas context"));
            }
        } catch (error) {
            reject(error);
        }
    });
};
