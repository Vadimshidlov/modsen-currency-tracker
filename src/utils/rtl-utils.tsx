import React, { PropsWithChildren } from "react";
import { Store, applyMiddleware, createStore } from "redux";
import { thunk } from "redux-thunk";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import { Provider } from "react-redux";
import { RootStateType, rootReducer } from "@/store/reducers";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
    preloadedState?: Partial<RootStateType>;
    store?: Store;
}

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        store = createStore(rootReducer, preloadedState, applyMiddleware(thunk)),
        ...renderOptions
    }: ExtendedRenderOptions = {},
) {
    function Wrapper({ children }: PropsWithChildren<object>): JSX.Element {
        return <Provider store={store}>{children}</Provider>;
    }
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export * from "@testing-library/react";

export { renderWithProviders as render };
