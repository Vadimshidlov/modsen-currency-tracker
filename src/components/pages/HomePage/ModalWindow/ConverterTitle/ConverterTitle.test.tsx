import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import { Provider } from "react-redux";
import { store } from "@/store";
import { renderWithProviders, screen } from "@/utils/rtl-utils";
import ConvertTitle from "@/components/pages/HomePage/ModalWindow/ConverterTitle/ConverterTitle";

describe("CurrencyStocks component tests", () => {
    it("should correct render component", async () => {
        renderWithProviders(<ConvertTitle currentCurrencyCode="USD" currentCurrencyValue={1.5} />);

        expect(screen.getByText("Commercial Dollar")).toBeInTheDocument();
        expect(screen.getByText("Value for USD:")).toBeInTheDocument();
        expect(screen.getByText("1.5")).toBeInTheDocument();
    });

    it("snapshot matches", async () => {
        const { asFragment } = renderWithProviders(
            <ConvertTitle currentCurrencyCode="USD" currentCurrencyValue={1.5} />,
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("should match snapshot", () => {
        const tree = renderer
            .create(
                <Provider store={store}>
                    <ConvertTitle currentCurrencyCode="USD" currentCurrencyValue={1.5} />
                </Provider>,
            )
            .toJSON();

        expect(tree).toMatchSnapshot();
    });
});
