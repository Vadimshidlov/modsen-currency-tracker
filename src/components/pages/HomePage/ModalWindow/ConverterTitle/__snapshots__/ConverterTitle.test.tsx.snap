// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyStocks component tests should match snapshot 1`] = `
<div
  className="modal__title"
>
  <div
    className="modal__selected-currency selected-currency"
  >
    <div
      className="selected-currency__title"
    >
      <svg />
      <p
        className=""
      >
        Commercial Dollar
      </p>
    </div>
    <div
      className="selected-currency__value"
    >
      <p
        className=""
      >
        Value for USD:
      </p>
      <p
        className=""
      >
        1.5
      </p>
    </div>
  </div>
</div>
`;

exports[`CurrencyStocks component tests snapshot matches 1`] = `
<DocumentFragment>
  <div
    class="modal__title"
  >
    <div
      class="modal__selected-currency selected-currency"
    >
      <div
        class="selected-currency__title"
      >
        <svg />
        <p
          class=""
        >
          Commercial Dollar
        </p>
      </div>
      <div
        class="selected-currency__value"
      >
        <p
          class=""
        >
          Value for USD:
        </p>
        <p
          class=""
        >
          1.5
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
